walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,224 @@
+    Integer endIndex = 0;
+
+    // We try to start the meeting whenever an event ends
+    for(Integer startIndex = 0; startIndex < eventPoints.size(); startIndex++) {
+      EventPoint startPoint = eventPoints.get(startIndex);
+      Event startEvent = startPoint.getEvent();
+
+      // If it's a starting point, just add the event to the intersectEvents array
+      // if it contains mandatory attendees
+      if(startPoint.getType() == EventPoint.Type.START) {
+        if(!areDisjoint(startEvent.getAttendees(), request.getAttendees())) {
+          intersectEvents.add(startEvent);
+        }
+        continue;
+      }
+
+      // It's an ending point, remove the event from the array
+      intersectEvents.remove(startEvent);
+
+      // If there is still at least an event with mandatory attendees during startPoint
+      if(!intersectEvents.isEmpty())
+        continue;
+
+      // We try to increase the end point as much as we can to make the range bigger
+      while(endIndex < eventPoints.size()) {
+        EventPoint endPoint = eventPoints.get(endIndex);
+
+        // We make sure the ending point is in the right of the starting point
+        // We also make sure our ending point is just before and event starts
+        if(endIndex > startIndex && endPoint.getType() == EventPoint.Type.START) {
+          TimeRange meetingRange = 
+          TimeRange.fromStartEnd(startPoint.getTime(), 
+                                 endPoint.getTime(), 
+                                 false);
+
+          // There are two cases when we don't want to increase the ending point anymore
+          // and we want to save the current range if it's long enough:
+          // 1. We reached the last point (the one we added at the end of the day)
+          // 2. The event that starts now contains mandatory attendees, so going further will
+          //    make it impossible for them to participate
+          if(endIndex == eventPoints.size() - 1 || 
+             !areDisjoint(endPoint.getAttendees(), request.getAttendees())) {
+            
+            // If the duration of the range is long enough
+            // and the ending point is not the same with the one we had at the previous starting point
+            if(meetingRange.duration() >= request.getDuration() &&
+               (answer.isEmpty() || meetingRange.end() > answer.get(answer.size()-1).end())) {
+              answer.add(meetingRange);
+            }
+            break;
+          }
+        }
+        endIndex++;
+      }
+    }
+
+    return answer;
+  }
+
+  /**
+   * Returns an intersection between two collections
+   */
+  private Collection<String> getIntersect(Collection<String> a, Collection<String> b) {
+    Collection<String> ret = new HashSet<String>();
+    for(String str : a) {
+      if(b.contains(str)) {
+        ret.add(str);
+      }
+    }
+    return ret;
+  }
+
+  private Boolean areDisjoint(Collection<String> a, Collection<String> b) {
```

Comment:
You can optimize it for both time and space without using getInterset function.

Reply:
Right, thank you! I changed it